### PR TITLE
fix(apiconvert): ToAPI for objects and arrays

### LINKED ISSUE #1088: [Kafka] ip_filter_object error when applying
## What happened?

I have these error when i'm applying change on ip_filter_object:

```
Error: error updating (member-streaming-platform-kafka-training-eu) service: 400: {"errors":[{"message":"Invalid 'user_config': Invalid input for ip_filter: 'network' is a required property","status":400}],"message":"Invalid 'user_config': Invalid input for ip_filter: 'network' is a required property"} -
with aiven_kafka.myservice
```

## What did you expect to happen?

I am trying to remove 1 ip of the ip_filter_object blocks 
Plan result: 
```
    ~ kafka_user_config {
            (8 unchanged attributes hidden)

          ~ ip_filter_object {
              ~ network = "aaa.aaa.aaa.a/23" -> "bbb.bbb.bbb.b/23"
            }
          ~ ip_filter_object {
              ~ network = "bbb.bbb.bbb.b/23" -> "ccc.ccc.c.c/17"
            }
          ~ ip_filter_object {
              ~ network = "ccc.ccc.c.c/17" -> "ddd.ddd.dd.d/23"
            }
          ~ ip_filter_object {
              ~ network = "ddd.ddd.dd.d/23" -> "eee.eee.ee.e/23"
            }
          ~ ip_filter_object {
              + description = "test"
              ~ network     = "eee.eee.ee.e/23" -> "fff.fff.ff.fff/32"
            }
          ~ ip_filter_object {
              - description = "test" -> null
              ~ network     = "fff.fff.ff.fff/32" -> "ggg.ggg.gg.g/23"
            }
          ~ ip_filter_object {
              ~ network = "ggg.ggg.gg.g/23" -> "hhh.hhh.hh.h/23"
            }
          ~ ip_filter_object {
              ~ network = "hhh.hhh.hh.h/23" -> "iii.iii.iii.i/24"
            }
          ~ ip_filter_object {
              ~ network = "iii.iii.iii.i/24" -> "jjj.jjj.jjj.j/18"
            }
          ~ ip_filter_object {
              ~ network = "jjj.jjj.jjj.j/18" -> "kkk.kkk.kkk.k/23"
            }
          ~ ip_filter_object {
              ~ network = "kkk.kkk.kkk.k/23" -> "lll.ll.lll.l/23"
            }
          - ip_filter_object {
              - network = "lll.ll.lll.l/23" -> null
            }

            (40 unchanged blocks hidden)
        }

        (1 unchanged block hidden)
    }
```

(I have overwrite all IPs by char. The ip that i delete is: aaa.aaa.aaa.a/23)
I suppose the other change is because all blocks are sorted.

## What else do we need to know?

Terraform code for ip_filter_object blocks:

```
    dynamic "ip_filter_object" {
      for_each = local.list_ip_filter_objects

      content {
        description = ip_filter_object.value.description
        network     = ip_filter_object.value.network
      }
    }
```

Terraform version: 1.3.9
Aiven provider version: 4.1.0

### PR DESCRIPTION
**Pull request by Serpentiel**
## About this change—what it does

fixes `ToAPI` for objects and arrays

resolves #1088

## Why this way

to make it work :)

### CODE REVIEW

**Comment by Serpentiel** [comment]
acceptance tests: https://github.com/aiven/terraform-provider-aiven/actions/runs/4429629269

File: internal/schemautil/userconfig/apiconvert/toapi.go [hunk 1]
```diff
@@ -176,10 +176,11 @@ func itemToAPI(
 	//
 	// We check that there are more than three elements in the fk slice, because we don't want to send the value if
 	// the parent object is the root object.
-	_, e := d.GetOk(fks)
-
-	if o && e && len(fk) > 3 && d.HasChange(fks[:strings.LastIndex(fks, ".0")]) {
-		o = false
+	if o && len(fk) > 3 {
+		match := regexp.MustCompile(`\.\d($|\.)`).FindAllStringIndex(fks, -1)
```
[the remark this answers is not in the document]
Comment (Serpentiel) [reply internal/schemautil/userconfig/apiconvert/toapi.go:181]: hmm, looks like we have multiple places in this file already where we do it this way, so it worths that we change it in a separate PR—I'll make one right after this one gets merged

**Review by ivan-savciuc — DISMISSED** [review]
These are results of my tests: 

1.  I have this user config:

```hlc
 kafka_user_config {
    ip_filter_object {
      network = "0.0.0.0/22"
    }

    ip_filter_object {
      network = "1.0.0.0/15"
    }

    ip_filter_object {
      network = "0.0.0.0/22"
    }

    ip_filter_object {
      network = "3.0.0.0/10"
    }

    ip_filter_object {
      network = "2.0.0.0/19"
    }
  }
```
   And if I snaffle some of the elements wit each other I get a massive diff without adding anything new:
```
 # aiven_kafka.k1 will be updated in-place
  ~ resource "aiven_kafka" "k1" {
        id                      = "aiven-ci-terraform-provider/ivans-k1"
        # (21 unchanged attributes hidden)

      ~ kafka_user_config {
            # (8 unchanged attributes hidden)

          ~ ip_filter_object {
              ~ network = "0.0.0.0/22" -> "3.0.0.0/10"
            }
          ~ ip_filter_object {
              ~ network = "1.0.0.0/15" -> "0.0.0.0/22"
            }
          ~ ip_filter_object {
              ~ network = "0.0.0.0/22" -> "1.0.0.0/15"
            }
          ~ ip_filter_object {
              ~ network = "3.0.0.0/10" -> "0.0.0.0/22"
            }

            # (3 unchanged blocks hidden)
        }
```
2. The removal of the elements is not smooth either, if we drop one element with the same user config I get this diff: 
```
 # aiven_kafka.k1 will be updated in-place
  ~ resource "aiven_kafka" "k1" {
        id                      = "aiven-ci-terraform-provider/ivans-k1"
        # (21 unchanged attributes hidden)

      ~ kafka_user_config {
            # (8 unchanged attributes hidden)

          ~ ip_filter_object {
              ~ network = "3.0.0.0/10" -> "2.0.0.0/19"
            }
          - ip_filter_object {
              - network = "2.0.0.0/19" -> null
            }

            # (5 unchanged blocks hidden)
        }

        # (1 unchanged block hidden)
    } 
```

**Comment by Serpentiel** [comment]
> And if I snaffle some of the elements wit each other I get a massive diff without adding anything new:

> The removal of the elements is not smooth either, if we drop one element with the same user config I get this diff:

I'm sorry but this works as expected and is not a part of this fix

File: internal/schemautil/userconfig/apiconvert/toapi.go [hunk 2]
```diff
@@ -176,10 +177,14 @@ func itemToAPI(
...
 	// the parent object is the root object.
-	_, e := d.GetOk(fks)
+	if o && len(fk) > 3 {
+		match := regexp.MustCompile(`\.\d($|\.)`).FindAllStringIndex(fks, -1)
 
-	if o && e && len(fk) > 3 && d.HasChange(fks[:strings.LastIndex(fks, ".0")]) {
-		o = false
+		_, e := d.GetOk(fks)
+
+		if match != nil && (e || !reflect.ValueOf(v).IsZero()) && d.HasChange(fks[:match[len(match)-1][0]]) {
```
Comment (ivan-savciuc) [inline internal/schemautil/userconfig/apiconvert/toapi.go:185]: We can benefit from having some of the comments, and it needs to be clarified at first glance what we are doing here.

**Review by ivan-savciuc — APPROVED** [review]
I cannot reproduce this error from API, LGTM

**Review by byashimov — DISMISSED** [review]

File: internal/schemautil/userconfig/apiconvert/toapi.go (hunk 2 above)
Comment (Serpentiel) [reply internal/schemautil/userconfig/apiconvert/toapi.go:185]: there's actually a comment already above this whole line, but sure we could make this clearer! I'll update the PR

**Review by byashimov — APPROVED** [review]

**Pull request merged**